packages/big-design/src/components/TableNext/Row/useRowState.ts

```diff
@@ -0,0 +1,68 @@
+import { TableSelectable } from '../types';
+
+interface UseRowStateProps<T> {
+  childRowIndex?: number;
+  childrenRows: T[];
+  isExpandable: boolean;
+  isParentRow: boolean;
+  isSelected?: boolean;
+  selectedItems: TableSelectable['selectedItems'];
+  onExpandedRow?(parentRowIndex: number | null): void;
+  onItemSelect?(
+    childRowIndex: number | null,
+    index: number,
+    childrenRows: T[],
+    isParentRow: boolean,
+    isExpandable: boolean,
+  ): void;
+  parentRowIndex: number;
+}
+
+export const useRowState = <T>({
+  childRowIndex,
+  childrenRows,
+  isExpandable,
+  isParentRow,
+  isSelected,
+  selectedItems,
+  onExpandedRow,
+  onItemSelect,
+  parentRowIndex,
+}: UseRowStateProps<T>) => {
+  const onChange = () => {
+    if (onItemSelect) {
+      onItemSelect(childRowIndex ?? null, parentRowIndex, childrenRows, isParentRow, isExpandable);
+    }
+  };
+
+  const onExpandedChange = () => {
+    if (onExpandedRow) {
+      onExpandedRow(parentRowIndex ?? null);
+    }
+  };
+
+  const hasChildrenRows = childrenRows.length > 0;
+
+  const allChildrenRowsSelected: boolean =
+    isExpandable &&
+    childrenRows.every((_childRow, childRowIndex) => {
```

Comment:
could create a helper function for this and pass in an argument for the 'filter' function. (`some`, `every`)  since the logic is the same for `allChildrenRowsSelected` and `someChildrenRowsSelected`

Reply:
Tried to create the helper function, but I think is now looking too bulky or distracting? 🤔  was there another way to implement it that I might be missing?

```ts
Could be outside the hook: 
const childrenRowsSelected = <T>(
  mode: 'every' | 'some',
  isExpandable: boolean,
  childrenRows: T[],
  parentRowIndex: number,
  selectedItems: TableSelectable['selectedItems'],
) => {
  if (!isExpandable) {
    return false;
  }

  if (mode === 'every') {
    return childrenRows.every((_childRow, childRowIndex) => {
      return selectedItems[`${parentRowIndex}.${childRowIndex}`] !== undefined;
    });
  }

  return childrenRows.some((_childRow, childRowIndex) => {
    return selectedItems[`${parentRowIndex}.${childRowIndex}`] !== undefined;
  });
};

or Inside the hook:
  const childrenRowsSelected = <T>(
    mode: 'every' | 'some',
  ) => {
    if (!isExpandable) {
      return false;
    }
  
    if (mode === 'every') {
      return childrenRows.every((_childRow, childRowIndex) => {
        return selectedItems[`${parentRowIndex}.${childRowIndex}`] !== undefined;
      });
    }
  
    return childrenRows.some((_childRow, childRowIndex) => {
      return selectedItems[`${parentRowIndex}.${childRowIndex}`] !== undefined;
    });
  };
```

Reply:
`const filterChildrenRows = (type) => {
	const dictionary = {
  	'some': Array.some,
    'every': Array.every,
    //extensible to other filters
  }
  const filterFunction = dictionary[type] ?? null;
  const cb = (el, childRowIndex, childrenRows) => selectedItems[`${parentRowIndex}.${childRowIndex}`] !== undefined
  
  return filterFunction ? filterFunction(cb) : false;
}

const allChildrenRowsSelected = isExpandable && filterChildrenRows('every')`

 was thinking something along these lines, but not sure if it will work 🤔

Reply:
I'd avoid using a dictionary to point to a prototype method. I'm actually more fond of leaving it as is since it's a bit more readable. If we did want to make it a little better, we could rip it outside of the hook and have it be a bit more generic. But I would only do that if we are going to reuse it in other places 🤷 
```ts
// "All" example
const areAllRowsSelected = (rows, selectedItems) => {
  return rows.some((row, index) => {
      const parentIndex = getParentRowIndex(row); // TODO: Define utility function;

      return selectedItems[`${parentIndex}.${index}`] !== undefined;
  });
}
```